### PR TITLE
Dynamic Index Name in OpenSearch sink  - Resolves #1459

### LINKED ISSUE #1459: Dynamic Index Name in OpenSearch sink
**Is your feature request related to a problem? Please describe.**

Some use-cases require placing different documents in different dynamic indices. Pipeline authors want to configure a dynamic index name that is derived from a property (or multiple properties) in a Data Prepper event.

**Describe the solution you'd like**

Support dynamic index names using a format string. This format string can use `${}` to signal string interpolation and use JSON Pointer for getting fields from events. For example: `metadata-${metadataType}`.

```
pipeline:
  ...
  sink:
    opensearch:
      hosts: ["https://opensearch-host"]
      index_type: custom
      index: "metadata-${metadataType}"
```

**Describe alternatives you've considered (Optional)**

Using conditional routing could allow for supporting a predefined number of indices. However, this approach has a very practical bound. Using this format string allows for unlimited (from Data Prepper's perspective) indices.

**Additional context**

N/A

### PR DESCRIPTION
**Pull request by kkondaka**
Signed-off-by: Krishna Kondaka <krishkdk@amazon.com>

### Description
Added support for dynamic index name in OpenSearch Sink. See issue #1459 (https://github.com/opensearch-project/data-prepper/issues/1459) for more details.
With this change, the `index`  option for Opensearch sink may contain dynamic parts "${..}" which are replaced by the corresponding value in the event that is being processed. 

Added new test cases for the new API in the Event. Added integration test case for testing dynamic index 

Resolves #1459 
 
### Issues Resolved
#1459
 
### Check List
- [ X] New functionality includes testing.
- [X ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [X ] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Comment by deepdatta** [comment]
Hi @kkondaka, the build failed because of missing javadocs in some of the new functions. Can you please address these.

File: data-prepper-api/src/main/java/org/opensearch/dataprepper/model/event/JacksonEvent.java [hunk 1]
```diff
@@ -254,6 +254,38 @@ public String toJsonString() {
...
+        String result = "";
+        int pos = 0;
+        while ((pos = format.indexOf("${", fromIndex)) != -1) {
+          int endPos = format.indexOf("}", pos+1);
+          if (endPos == -1) {
+            throw new RuntimeException("index name not properly formed");
+          }
+          result += format.substring(fromIndex, pos);
+          String name = format.substring(pos+2, endPos);
+          Object val = this.get(name, String.class);
```
Comment (dlvenable) [inline data-prepper-api/src/main/java/org/opensearch/dataprepper/model/event/JacksonEvent.java:276]: This value may not always be a `String` class. I think it would be best to:

```
Object value = this.get(name, Object.class);
```

And then below on line 280:

```
result += value.toString();
```

File: data-prepper-api/src/main/java/org/opensearch/dataprepper/model/event/JacksonEvent.java [hunk 2]
```diff
@@ -254,6 +254,38 @@ public String toJsonString() {
...
+     * string may contain parts with format "${.../.../...}" which are replaced
+     * by their value in the event
+     * @param format string with format
+     * @throws RuntimeException if the format is incorrect or the value is not a string
+     */
+    @Override
+    public String formatString(final String format) {
+        int fromIndex = 0;
+        String result = "";
+        int pos = 0;
```
Comment (dlvenable) [inline data-prepper-api/src/main/java/org/opensearch/dataprepper/model/event/JacksonEvent.java:268]: Please prefer full works for variable names. It's not too big of a deal with variables within functions. But, since I have some comments here anyway it might be nice to clean these up.

File: data-prepper-api/src/test/java/org/opensearch/dataprepper/model/event/JacksonEventTest.java [hunk 3]
```diff
@@ -456,6 +456,33 @@ public void testBuild_withStringData() {
...
+                .withEventType(eventType)
+                .withData(jsonString)
+                .getThis()
+                .build();
+
+        assertThat(event.formatString("test-${foo}-string"), is(equalTo("test-bar-string")));
+        assertThat(event.formatString("test-string-${foo}"), is(equalTo("test-string-bar")));
+        assertThat(event.formatString("${foo}-test-string"), is(equalTo("bar-test-string")));
+        assertThat(event.formatString("test-${boo}-string"), is(equalTo(null)));
+        assertThrows(RuntimeException.class, () -> event.formatString("test-${foo-string"));
```
Comment (dlvenable) [inline data-prepper-api/src/test/java/org/opensearch/dataprepper/model/event/JacksonEventTest.java:474]: This should be it's own test case. That is, it would have its own function.

File: data-prepper-api/src/test/java/org/opensearch/dataprepper/model/event/JacksonEventTest.java [hunk 4]
```diff
@@ -456,6 +456,33 @@ public void testBuild_withStringData() {
...
+
+        final String jsonString = "{\"foo\": \"bar\", \"info\": {\"ids\": {\"id\":\"idx\"}}}";
+
+        event = JacksonEvent.builder()
+                .withEventType(eventType)
+                .withData(jsonString)
+                .getThis()
+                .build();
+
+        assertThat(event.formatString("test-${foo}-string"), is(equalTo("test-bar-string")));
```
Comment (dlvenable) [inline data-prepper-api/src/test/java/org/opensearch/dataprepper/model/event/JacksonEventTest.java:470]: Ideally these different tests are not in one test case. You can make all the success tests a parameterized test and it would work quite nicely.

You can use JUnit's `@CsvSource` to separate the actual value from expected. See this documentation: https://junit.org/junit5/docs/5.8.1/api/org.junit.jupiter.params/org/junit/jupiter/params/provider/CsvSource.html#value().

You would not include the negative case with the exception.

File: data-prepper-api/src/test/java/org/opensearch/dataprepper/model/event/JacksonEventTest.java [hunk 5]
```diff
@@ -456,6 +456,33 @@ public void testBuild_withStringData() {
         assertThat(event.get("foo", String.class), is(equalTo("bar")));
     }
 
+    @Test
```
Comment (dlvenable) [inline data-prepper-api/src/test/java/org/opensearch/dataprepper/model/event/JacksonEventTest.java:459]: Please also add some test cases with non-String values in the Event. You will probably find that these fail because of the code which casts strings. See my comment above.

File: data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/OpenSearchSink.java [hunk 6]
```diff
@@ -98,17 +103,28 @@ public OpenSearchSink(final PluginSetting pluginSetting) {
   public void initialize() throws IOException {
     LOG.info("Initializing OpenSearch sink");
     restHighLevelClient = openSearchSinkConfig.getConnectionConfiguration().createClient();
-    indexManager = indexManagerFactory.getIndexManager(indexType, restHighLevelClient, openSearchSinkConfig);
+    configuredIndexAlias = openSearchSinkConfig.getIndexConfiguration().getIndexAlias();
     final String dlqFile = openSearchSinkConfig.getRetryConfiguration().getDlqFile();
     if (dlqFile != null) {
       dlqWriter = Files.newBufferedWriter(Paths.get(dlqFile), StandardOpenOption.CREATE, StandardOpenOption.APPEND);
     }
+    if (!isDynamicIndexAlias(configuredIndexAlias)) {
```
Comment (dlvenable) [inline data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/OpenSearchSink.java:111]: I'm a little concerned about leaking management of the index into this class. I think that we could handle the decisions on dynamic vs. static index management in the `IndexManagerFactory`. This would reduce the number of changes in this class.

File: data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/OpenSearchSink.java [hunk 7]
```diff
@@ -58,20 +60,22 @@ public class OpenSearchSink extends AbstractSink<Record<Event>> {
...
   private final String routingField;
   private final String action;
+  private String configuredIndexAlias;
 
   private final Timer bulkRequestTimer;
   private final Counter bulkRequestErrorsCounter;
   private final DistributionSummary bulkRequestSizeBytesSummary;
   private OpenSearchClient openSearchClient;
   private ObjectMapper objectMapper;
+  private Map<String, IndexManager> indexManagersMap;
```
Comment (dlvenable) [inline data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/OpenSearchSink.java:78]: This should be a cache. I could image users using values here that change over time such that this map would grow indefinitely.

You can look into Guava for some caches in the `com.google.common.cache` package. Also, there is the Caffeine project. But, I don't think Data Prepper current has that in the final class path.

Also, I think this cache should be moved out of this class as part of the refactoring I suggested below.

File: data-prepper-plugins/opensearch/src/integrationTest/java/org/opensearch/dataprepper/plugins/sink/opensearch/OpenSearchSinkIT.java [hunk 8]
```diff
@@ -584,6 +585,34 @@ public void testOpenSearchRoutingField(final String testRoutingField) throws IOE
     sink.shutdown();
   }
 
+  @ParameterizedTest
+  @ValueSource(strings = {"info/ids/id", "id"})
+  public void testOpenSearchDynamicIndex(final String testIndex) throws IOException, InterruptedException {
```
Comment (dlvenable) [inline data-prepper-plugins/opensearch/src/integrationTest/java/org/opensearch/dataprepper/plugins/sink/opensearch/OpenSearchSinkIT.java:590]: We should also have an integration test that combines a dynamic index based on properties (as you have here) with a time-based index names (e.g. `logs-%{yyyy.MM}`.

Relatedly, I don't see an integration test which actually verifies time-based index names. We should have a test for that in isolation to ensure it is not affected by this change.

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: data-prepper-api/src/test/java/org/opensearch/dataprepper/model/event/JacksonEventTest.java [hunk 9]
```diff
@@ -456,6 +456,33 @@ public void testBuild_withStringData() {
...
+    public void testBuild_withFormatString() {
```
Comment (dlvenable) [inline data-prepper-api/src/test/java/org/opensearch/dataprepper/model/event/JacksonEventTest.java:460]: Also, there should be some test cases for other edge cases - The value is `null`. The value is an empty string.

Comment (kkondaka) [reply data-prepper-api/src/test/java/org/opensearch/dataprepper/model/event/JacksonEventTest.java:460]: I have added the edge test cases and it looks like all tests are passing now.

File: data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/index/DynamicIndexManager.java [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+	this.indexType = indexType;
+        this.indexManagerFactory = new IndexManagerFactory();
+        this.restHighLevelClient = restHighLevelClient;
+        this.openSearchSinkConfiguration = openSearchSinkConfiguration;
+        CacheBuilder<String, IndexManager> cacheBuilder = CacheBuilder.newBuilder()
+	                .recordStats()
+		        .concurrencyLevel(1)
+		        .maximumWeight(cacheSizeInKB)
+			.expireAfterAccess(CACHE_EXPIRE_AFTER_ACCESS_TIME_MINUTES, TimeUnit.MINUTES)
+	                .weigher((k, v) -> 32);
```
Comment (dlvenable) [inline data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/index/DynamicIndexManager.java:64]: Ideally this is the name of the index plus the `IndexManager` which is associated with it.

Conceptually this would be:

```
(k, v) -> k.length() + computeSize(v)
```

But, I think it is unnecessary to compute that size for now. So using `32` here is a good approach. But, please make this `32` a named constant so that it is not some magic number.

Thus this line should become:

```
.weigher((k, v) -> k.length() + APPROXIMATE_INDEX_MANAGER_SIZE);
```

File: data-prepper-plugins/opensearch/src/test/java/org/opensearch/dataprepper/plugins/sink/opensearch/index/DynamicIndexManagerTests.java [hunk 11]
```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.dataprepper.plugins.sink.opensearch.index;
+
+import org.junit.After;
```
Comment (dlvenable) [inline data-prepper-plugins/opensearch/src/test/java/org/opensearch/dataprepper/plugins/sink/opensearch/index/DynamicIndexManagerTests.java:8]: Please update this test to use JUnit 5. All new tests should be using JUnit 5.

File: data-prepper-plugins/opensearch/src/test/java/org/opensearch/dataprepper/plugins/sink/opensearch/index/DynamicIndexManagerTests.java [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+import static org.junit.Assert.assertThrows;
+import static org.junit.Assert.assertTrue;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoMoreInteractions;
+import static org.mockito.Mockito.when;
+import static org.mockito.MockitoAnnotations.initMocks;
+
+public class DynamicIndexManagerTests {
```
Comment (dlvenable) [inline data-prepper-plugins/opensearch/src/test/java/org/opensearch/dataprepper/plugins/sink/opensearch/index/DynamicIndexManagerTests.java:49]: Please add a test which verifies the caching behavior.

File: data-prepper-plugins/opensearch/src/test/java/org/opensearch/dataprepper/plugins/sink/opensearch/index/DynamicIndexManagerTests.java [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+        when(restHighLevelClient.indices()).thenReturn(indicesClient);
+    }
+
+    @Test
+    public void dynamicIndexBasicTest(){
+        when(indexConfiguration.getIndexAlias()).thenReturn(INDEX_ALIAS);
+        when(clusterGetSettingsResponse.getSetting(IndexConstants.ISM_ENABLED_SETTING)).thenReturn("true");
+        String configuredIndexAlias = openSearchSinkConfiguration.getIndexConfiguration().getIndexAlias();
+	try {
+            indexManager = indexManagerFactory.getIndexManager(IndexType.CUSTOM, restHighLevelClient, openSearchSinkConfiguration, configuredIndexAlias);
```
Comment (dlvenable) [inline data-prepper-plugins/opensearch/src/test/java/org/opensearch/dataprepper/plugins/sink/opensearch/index/DynamicIndexManagerTests.java:111]: This test case is mixing two different behaviors.

* Verify the behavior of `getIndexManager`
* Verifying the behavior of `DynamicIndexManager.getIndexName`

Test cases should only verify a single behavior. Please verify `IndexManagerFactory.getIndexManager` in `IndexManagerFactoryTests`.

File: data-prepper-plugins/opensearch/src/test/java/org/opensearch/dataprepper/plugins/sink/opensearch/index/DynamicIndexManagerTests.java [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+    @Mock
+    private ResponseException responseException;
+
+    static final String EVENT_TYPE = "event";
+
+    @Before
+    public void setup() throws IOException {
+        initMocks(this);
+
+        indexManagerFactory = new IndexManagerFactory();
```
Comment (dlvenable) [inline data-prepper-plugins/opensearch/src/test/java/org/opensearch/dataprepper/plugins/sink/opensearch/index/DynamicIndexManagerTests.java:95]: This should be a mocked object. It should help you write the tests I suggested above.

Also, see my other comment about testing only one behavior at a time. Once you split the tests you won't need a real object here.

File: data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/index/IndexManager.java [hunk 15]
```diff
@@ -1,286 +1,9 @@
...
-            ((Map<String, Object>) configuration.getIndexTemplate().get("settings"))
-                    .put(IndexConstants.ISM_POLICY_ID_SETTING, ismPolicyId);
-        }
-        ((Map<String, Object>) configuration.getIndexTemplate().get("settings"))
-                .put(IndexConstants.ISM_ROLLOVER_ALIAS_SETTING, rolloverAlias);
-    }
 
+public interface IndexManager{
+	void setupIndex() throws IOException;
+	String getIndexName(final String indexAlias) throws IOException;
```
Comment (dlvenable) [inline data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/index/IndexManager.java:7]: Why did you add two overloads for `getIndexName()`? This could be somewhat confusing because it is unclear which method a caller (ie. `OpenSearchSink`) should use.

File: data-prepper-api/src/test/java/org/opensearch/dataprepper/model/event/JacksonEventTest.java [hunk 16]
```diff
@@ -456,6 +457,74 @@ public void testBuild_withStringData() {
         assertThat(event.get("foo", String.class), is(equalTo("bar")));
     }
 
+    @ParameterizedTest
+    @CsvSource({
+        "test-${foo}-string, test-bar-string",
```
Comment (dlvenable) [inline data-prepper-api/src/test/java/org/opensearch/dataprepper/model/event/JacksonEventTest.java:462]: One other test case we should have - a string without any formatting. It should return the input value.

```
"test-string, test-string"
```

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/index/IndexManager.java (hunk 15 above)
Comment (kkondaka) [reply data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/index/IndexManager.java:7]: @dlvenable, I thought it is easier to use this method when it is known that the indexAlias is static one. Also, this is equivalent to default behavior of getIndexName() where the index alias is taken from the config. Do you prefer to not have this? Or make it Optional?

File: data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/index/IndexManagerFactory.java [hunk 17]
```diff
@@ -8,32 +8,59 @@
...
+	    return null;
+	}
+    }
+
+    public final IndexManager getIndexManager(final IndexType indexType,
+                                        final RestHighLevelClient restHighLevelClient,
+                                        final OpenSearchSinkConfiguration openSearchSinkConfiguration,
+					String indexAlias) throws IOException {
+        if (indexAlias != null && isDynamicIndexAlias(indexAlias)) {
+	    return  new DynamicIndexManager(indexType, restHighLevelClient, openSearchSinkConfiguration);
```
Comment (dlvenable) [inline data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/index/IndexManagerFactory.java:31]: You can add a parameter to the `DynamicIndexManager` constructor which takes an `IndexManagerFactory`. Then you can pass `this` in here.

```
return new DynamicIndexManager(indexType, restHighLevelClient, openSearchSinkConfiguration, this);
```

File: data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/index/IndexManager.java (hunk 15 above)
Comment (dlvenable) [reply data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/index/IndexManager.java:7]: The reason is that the `getIndexName(String)` is the only method exposed to external clients. So from an external perspective it is unclear which to call and what the results will be.

Comment (kkondaka) [reply data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/index/IndexManager.java:7]: Removed the getIndexName() API.

File: data-prepper-plugins/opensearch/src/test/java/org/opensearch/dataprepper/plugins/sink/opensearch/index/DynamicIndexManagerTests.java (hunk 11 above)
Comment (kkondaka) [reply data-prepper-plugins/opensearch/src/test/java/org/opensearch/dataprepper/plugins/sink/opensearch/index/DynamicIndexManagerTests.java:8]: I think we are already using junit5. I think the name is still org.junit.xxx.

File: data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/index/DynamicIndexManager.java [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+                        .weigher((k, v) -> APPROXIMATE_INDEX_MANAGER_SIZE);
+        this.indexManagerCache = cacheBuilder.build();
+        checkNotNull(indexManagerCache);
+    }
+
+    @Override
+    public void setupIndex() throws IOException {
+    }
+
+    public boolean isInCache(final String indexAlias) {
```
Comment (dlvenable) [inline data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/index/DynamicIndexManager.java:74]: This is unused and can be removed.

File: data-prepper-plugins/opensearch/src/test/java/org/opensearch/dataprepper/plugins/sink/opensearch/index/DynamicIndexManagerTests.java [hunk 19]
```diff
@@ -0,0 +1,183 @@
...
+        when(restHighLevelClient.indices()).thenReturn(indicesClient);
+        dynamicIndexManager = new DynamicIndexManager(IndexType.CUSTOM, restHighLevelClient, openSearchSinkConfiguration, mockIndexManagerFactory);
+    }
+
+    @Test
+    public void dynamicIndexBasicTest(){
+        when(indexConfiguration.getIndexAlias()).thenReturn(INDEX_ALIAS);
+        when(clusterGetSettingsResponse.getSetting(IndexConstants.ISM_ENABLED_SETTING)).thenReturn("true");
+        String configuredIndexAlias = openSearchSinkConfiguration.getIndexConfiguration().getIndexAlias();
+        try {
```
Comment (dlvenable) [inline data-prepper-plugins/opensearch/src/test/java/org/opensearch/dataprepper/plugins/sink/opensearch/index/DynamicIndexManagerTests.java:126]: Please remove these try-catch blocks (in lots of these tests).

You can add a throws clause to these test methods instead.

```
@Test
public void dynamicIndexBasicTest() throws IOException {
```

File: data-prepper-plugins/opensearch/src/test/java/org/opensearch/dataprepper/plugins/sink/opensearch/index/IndexManagerFactoryTests.java [hunk 20]
```diff
@@ -67,4 +70,14 @@ public void getIndexManager_returns_IndexManager_when_provided_MANAGEMENT_DISABL
         assertThat(indexManager, instanceOf(IndexManager.class));
     }
 
+    @Test
+    public void getIndexManager_dynamic_Default() {
+        when(indexConfiguration.getIndexAlias()).thenReturn(DYNAMIC_INDEX_ALIAS);
+        try {
```
Comment (dlvenable) [inline data-prepper-plugins/opensearch/src/test/java/org/opensearch/dataprepper/plugins/sink/opensearch/index/IndexManagerFactoryTests.java:76]: Remove try-catch.

File: data-prepper-plugins/opensearch/src/test/java/org/opensearch/dataprepper/plugins/sink/opensearch/index/DefaultIndexManagerTests.java [hunk 21]
```diff
@@ -100,8 +100,10 @@ public void setup() throws IOException {
     public void getIndexAlias_IndexWithTimePattern(){
         when(indexConfiguration.getIndexAlias()).thenReturn(INDEX_ALIAS_WITH_TIME_PATTERN);
         defaultIndexManager = indexManagerFactory.getIndexManager(IndexType.CUSTOM, restHighLevelClient, openSearchSinkConfiguration);
-        final String indexAlias = defaultIndexManager.getIndexAlias();
-        assertTrue(EXPECTED_INDEX_PATTERN.matcher(indexAlias).matches());
+        try {
+            final String indexAlias = defaultIndexManager.getIndexName(null);
+            assertTrue(EXPECTED_INDEX_PATTERN.matcher(indexAlias).matches());
```
Comment (dlvenable) [inline data-prepper-plugins/opensearch/src/test/java/org/opensearch/dataprepper/plugins/sink/opensearch/index/DefaultIndexManagerTests.java:105]: Please use `assertThat(string, matchPattern(pattern))`. You use this in another test in this PR.

**Review by dlvenable — CHANGES_REQUESTED** [review]

**Review by dlvenable — APPROVED** [review]

**Review by dlvenable — APPROVED** [review]
